src/mpid/ch4/src/ch4_progress.c

```diff
@@ -69,11 +83,14 @@ static int progress_test(MPID_Progress_state * state, int wait)
 #endif
 
     if (state->flag & MPIDI_PROGRESS_NM) {
-        if (!wait || !progress_made(state)) {
+        if (!wait || !state->progress_made) {
             if (do_global_progress()) {
                 for (int vci = 0; vci < MPIDI_global.n_vcis; vci++) {
                     MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(vci).lock);
                     mpi_errno = MPIDI_NM_progress(vci, 0);
+                    if (wait) {
+                        check_progress_made_vci(state, vci);
+                    }
```

Comment:
I'm feeling this kind of mutex protection seems not comfortable to me.

Every state has a `progress_made` variable, associated with all vcis in this state.
However, when mpi wants to change its value, it only locks a single vci lock.

Example is two threads are changing same state currently, using different vci lock.

I prefer just to make `progress_made` be an atomic value.

Reply:
The state is designed to be thread-safe -- no two threads will be accessing the same state or it is a severe code error.

Reply:
It is fine under that constraint. I just don't kind N to 1 mapping for mutexes.